File: updraft-sdk/src/main/java/com/appswithlove/updraft/feedback/form/FeedbackFormPresenter.java
package com.appswithlove.updraft.feedback.form;

import android.os.Handler;
import com.appswithlove.updraft.Updraft;
import com.appswithlove.updraft.api.ApiWrapper;
import com.appswithlove.updraft.feedback.FeedbackActivity;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

import java.util.concurrent.TimeUnit;

public class FeedbackFormPresenter {

    private FeedbackFormContract.View mView;
    private CompositeDisposable mCompositeDisposable;
    private ApiWrapper mApiWrapper;
    private Disposable mSendFeedbackDisposable;

    FeedbackFormPresenter() {
        mApiWrapper = Updraft.getInstance().getApiWrapper();
    }

    public void attachView(FeedbackFormContract.View view) {
        mView = view;
        mCompositeDisposable = new CompositeDisposable();
    }

    public void detachView() {
        mCompositeDisposable.dispose();
        mView = null;
    }

    public void onSendButtonClicked() {
        mView.showProgress();
        mSendFeedbackDisposable =
                mApiWrapper
                        .sendMobileFeedback(
                                mView.getSelectedChoice(),
                                mView.getDescription(),
                                mView.getEmail(),
                                FeedbackActivity.SAVED_SCREENSHOT
                        )
                        //.concatMap(progress -> Observable.just(progress).delay(1000, TimeUnit.MILLISECONDS))
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(
                                response -> mView.updateProgress(response),
                                t -> {
                                    t.printStackTrace();
                                    mView.showErrorMessage(t);
                                },
                                () -> {
                                    mView.showSuccessMessage();
                                    closeAfterDelay();
                                }
                        );
        mCompositeDisposable.add(mSendFeedbackDisposable);
    }

    public void onProgressCancelClicked() {
        if (mSendFeedbackDisposable != null && !mSendFeedbackDisposable.isDisposed()) {
            mSendFeedbackDisposable.dispose();
        }
        mView.hideProgress();
    }

    private void closeAfterDelay(){
        Handler handler = new Handler();
        handler.postDelayed(
                () -> mView.closeFeedback(),
                1000
        );
    }
}

File: updraft-sdk/src/main/java/com/appswithlove/updraft/feedback/form/FeedbackFormPresenter.kt
package com.appswithlove.updraft.feedback.form

import com.appswithlove.updraft.Updraft
import com.appswithlove.updraft.api.ApiWrapper
import com.appswithlove.updraft.feedback.FeedbackActivity
import kotlinx.coroutines.*
import kotlin.coroutines.CoroutineContext

class FeedbackFormPresenter : CoroutineScope {

    private var view: FeedbackFormContract.View? = null
    private val apiWrapper: ApiWrapper? = Updraft.getInstance()?.apiWrapper

    private var presenterJob: Job = SupervisorJob()
    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main + presenterJob

    private var sendJob: Job? = null

    fun attachView(view: FeedbackFormContract.View) {
        this.view = view
    }

    fun detachView() {
        cancel()
        view = null
    }

    fun onSendButtonClicked() {
        val currentView = view ?: return
        val api = apiWrapper ?: return

        currentView.showProgress()

        sendJob?.cancel()
        sendJob = launch {
            try {
                withContext(Dispatchers.IO) {
                    api.sendMobileFeedback(
                        currentView.getSelectedChoice(),
                        currentView.getDescription(),
                        currentView.getEmail(),
                        FeedbackActivity.SAVED_SCREENSHOT
                    ).collect { progress ->
                        withContext(Dispatchers.Main) {
                            currentView.updateProgress(progress)
                        }
                    }
                }

                withContext(Dispatchers.Main) {
                    currentView.showSuccessMessage()
                    closeAfterDelay()
                }
            } catch (_: CancellationException) {
                // Job was cancelled, ignore
            } catch (t: Throwable) {
                t.printStackTrace()
                withContext(Dispatchers.Main) {
                    currentView.showErrorMessage(t)
                }
            }
        }
    }

    fun onProgressCancelClicked() {
        sendJob?.cancel()
        view?.hideProgress()
    }

    private fun closeAfterDelay() {
        launch {
            delay(1000)
            view?.closeFeedback()
        }
    }
}
